R/methods-MbioDataset.R

```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
+#' getMetadataVariableNames(microbiomeData::DiabImmune)
+#' variablesDT <- getVariables(
+#'      microbiomeData::DiabImmune, 
+#'      list("metadata" = c("age_months", "sex"),
+#'           "16S (V4) Genus" = "Bacteroides", 
+#'           "WGS Metagenome enzyme pathway abundance data" = "ANAGLYCOLYSIS-PWY: glycolysis III (from glucose)"
+#'      )
+#' )
+#' @param object A Microbiome Dataset
+#' @param variables The names of the variables to return. this should be a named list
+#' where the names are collection names and the values are variable names for that collection.
+#' For the case of metadata variables, the name should be "metadata".
+#' @return a data.table of the requested variables
+#' @rdname getVariables
+#' @export
+setGeneric("getVariables", function(object, variables) standardGeneric("getVariables"), signature = "object")
+
+#' @rdname getVariables
+#' @aliases getVariables,MbioDataset,character-method
+setMethod("getVariables", "MbioDataset", function(object, variables) {
+
+    if (!is.list(variables)) {
+        stop("variables argument must be a list")
+    }
+    if (is.null(names(variables))) {
+        stop("variables argument must be a named list")
+    }
+
+    ## identify variables w identical names early
+    flattenedVars <- unlist(variables)
+    dups <- unname(flattenedVars[duplicated(flattenedVars)])
+    collectionsWithDups <- lapply(variables, function(x) {dups %in% x})
+    collectionsWithDupsIndexes <- unname(which(collectionsWithDups == TRUE))
+
+    fetchCollectionVariables <- function(collectionIndex) {
+        variableNames <- variables[[collectionIndex]]
+        collectionName <- names(variables)[collectionIndex]
+
+        if (collectionName == "metadata") {
+            return(getSampleMetadata(object, metadataVariables = variableNames))
+        }
+
+        if (!collectionName %in% getCollectionNames(object)) {
+            stop(sprintf("Collection '%s' does not exist", collectionName))
+        }
+
+        if (any(variableNames %in% getCollectionVariableNames(object, collectionName))) {
+            collection <- getCollection(object, collectionName)
+            presentVars <- variableNames[variableNames %in% getCollectionVariableNames(collection)]
+            if (veupathUtils::isOneToManyWithAncestor(collection)) {
+                warning("Unable to return the following variables: ", presentVars)
+                return(data.table::data.table())
+            }
+            dt <- veupathUtils::getCollectionData(collection, presentVars)
+            if (collectionIndex %in% collectionsWithDupsIndexes) {
+                ## rename variables to prepend the collection name
+                names(dt)[names(dt) %in% presentVars] <- paste(collectionName, presentVars)
+            }
+            return(dt)
+        } else {
+            stop(sprintf("Collection '%s' does not contain the following variables: %s", collectionName, paste(variableNames, collapse = ", ")))
```

Comment:
Should have different verbiage if the input was for the 'metadata'?

Reply:
metadata was returned before now..